akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala

```diff
@@ -0,0 +1,307 @@
+                settings.terminationDeadlineExceededResponse
+              } else grab(fromUser)
+
+            pendingUserHandlerResponse = false
+            push(toNet, response)
+          }
+        })
+
+        // once termination deadline hits, we stop pulling from network
+        setHandler(toUser, new OutHandler {
+          override def onPull(): Unit = {
+            // if (deadline.hasTimeLeft()) // we pull always as we want to reply errors to everyone
+            pull(fromNet)
+          }
+        })
+
+        setHandler(fromNet, new InHandler {
+          override def onPush(): Unit = {
+            val request = grab(fromNet)
+            log.warning(
+              "Terminating server ({}), attempting to send termination reply to incoming [{} {}]",
+              formatTimeLeft(deadline), request.method, request.uri.path)
+
+            // on purpose discard all incoming bytes for requests
+            // could discard with the deadline.timeLeft completion timeout, but not necessarily needed
+            request.entity.discardBytes()(interpreter.subFusingMaterializer).future.onComplete {
+              case Success(_) ⇒ // ignore
+              case Failure(ex) ⇒
+                // we do want to cause this failure to fail the termination eagerly
+                failureCallback.invoke(ex)
+            }(interpreter.materializer.executionContext)
+
+            // we can reply right away with an termination response since user handler will never emit a response anymore
+            push(toNet, settings.terminationDeadlineExceededResponse)
+            pull(fromNet)
+          }
+        })
+
+        // we continue pulling from user, to make sure we'd get the "final user reply" that may be sent during termination
+        setHandler(toNet, new OutHandler {
+          override def onPull(): Unit = {
+            if (pendingUserHandlerResponse) {
+              if (isAvailable(fromUser)) pull(fromUser)
+            } else if (pendingTerminationResponse) {
+              pendingTerminationResponse = false
+              push(toNet, settings.terminationDeadlineExceededResponse)
+            }
+          }
+        })
+      }
+
+      override def postStop(): Unit = {
+        terminationOfConnectionDone.success(Http.HttpConnectionTerminated)
+      }
+
+      override protected def onTimer(timerKey: Any): Unit = timerKey match {
+        case TerminationDeadlineTimerKey ⇒
+          val ex = new ServerTerminationDeadlineReached
+          if (pendingUserHandlerResponse) {
+            // sending the reply here is a "nice to try", but the stage failure will likely overtake it and terminate the connection first
+            emit(toNet, settings.terminationDeadlineExceededResponse, () ⇒ failStage(ex))
+          } else {
+            failStage(ex) // TODO think if to throw or fail here? this causes the "dont keep connections around longer than the deadline"
```

Comment:
thinking done?

Reply:
Yes, thanks for skimming TODO/FIXMEs, seems I missed some

Reply:
(It is correct to fail stage here)